src/analysis.ts

```diff
@@ -36,10 +39,55 @@ export class ValidationError extends Error {
 
 
 export class Analysis {
-  descriptors: DocumentDescriptor[];
+  private _descriptors: DocumentDescriptor[];
+  private _elementsByTagName = new Map<string, ElementDescriptor>();
+  private _elementsByPackageDir = new Map<string, ElementDescriptor[]>();
+  elementPaths = new Map<ElementDescriptor, string>();
+  locationOffsets = new Map<ElementDescriptor, LocationOffset>();
```

Comment:
Yes, that was the goal even as I wrote this, I just needed to control the complexity..